src/internal/clusterstate/v2.6.0/pfsdb.go

```diff
@@ -0,0 +1,580 @@
+	result := make([]*pfs.Commit, 0)
+	var c *pfs.Commit
+	for len(cs) > 0 {
+		c, cs = cs[0], cs[1:]
+		if c == nil {
+			continue
+		}
+		ci, ok := skipSet[oldCommitKey(c)]
+		if ok {
+			cs = append(cs, traverse(ci)...)
+		} else {
+			result = append(result, proto.Clone(c).(*pfs.Commit))
+		}
+	}
+	return result
+}
+
+func commitSubvenance(ctx context.Context, tx *pachsql.Tx, c *pfs.Commit) ([]*pfs.Commit, error) {
+	key := oldCommitKey(c)
+	query := `SELECT commit_id, branch FROM pfs.commits WHERE int_id IN (       
+            SELECT to_id FROM pfs.commits JOIN pfs.commit_provenance ON int_id = to_id WHERE commit_id = $1
+        );`
+	rows, err := tx.QueryxContext(ctx, query, key)
+	if err != nil {
+		return nil, errors.Wrapf(err, "query commit subvenance for commit %q", c)
+	}
+	commitSubvenance := make([]*pfs.Commit, 0)
+	for rows.Next() {
+		var commitId, branch string
+		if err := rows.Scan(&commitId, &branch); err != nil {
+			return nil, errors.EnsureStack(err)
+		}
+		project, rest, _ := strings.Cut(commitId, "/")
+		repoName, rest, _ := strings.Cut(rest, ".")
+		repoType, rest, _ := strings.Cut(rest, "@")
+		branch, id, _ := strings.Cut(rest, "=")
+		repo := &pfs.Repo{
+			Name: repoName,
+			Type: repoType,
+			Project: &pfs.Project{
+				Name: project,
+			},
+		}
+		c := &pfs.Commit{
+			Repo:   repo,
+			Branch: &pfs.Branch{Name: branch, Repo: repo},
+			ID:     id,
+		}
+		commitSubvenance = append(commitSubvenance, c)
+	}
+	return commitSubvenance, nil
+}
+
+// since the provenance relationship points from downstream to upstream, 'from' represents the commit subvenant to the 'to' commit
+func deleteCommitProvenance(ctx context.Context, tx *pachsql.Tx, from, to *pfs.Commit) error {
+	fromId, err := getCommitIntId(ctx, tx, from)
+	if err != nil {
+		return errors.Wrapf(err, "get commit int id for 'from' commit %q", from)
+	}
+	toId, err := getCommitIntId(ctx, tx, to)
+	if err != nil {
+		return errors.Wrapf(err, "get commit int id for 'to' commit %q", to)
+	}
+	_, err = tx.ExecContext(ctx, `DELETE FROM pfs.commit_provenance WHERE from_id=$1 AND to_id=$2;`, fromId, toId)
+	return errors.EnsureStack(err)
+}
+
+// since the provenance relationship points from downstream to upstream, 'from' represents the commit subvenant to the 'to' commit
+func addCommitProvenance(ctx context.Context, tx *pachsql.Tx, from, to *pfs.Commit) error {
+	fromId, err := getCommitIntId(ctx, tx, from)
+	if err != nil {
+		return err
+	}
+	toId, err := getCommitIntId(ctx, tx, to)
+	if err != nil {
+		return err
+	}
+	_, err = tx.ExecContext(ctx, `INSERT INTO pfs.commit_provenance(from_id, to_id) VALUES ($1, $2) ON CONFLICT DO NOTHING;`, fromId, toId)
+	return errors.EnsureStack(err)
+}
+
+// the goal of this migration is to migrate commits to be unqiuely indentified by a (repo, UUID), instead of by (repo, branch, UUID)
+//
+// 1) commits with equal (repo, UUID) pairs must be de-duplicated. This is primarily expected to happen in deferred processing cases.
```

Comment:
same as above, I would just reword this in terms of triggers (i.e. the triggered branch and the triggering branch).